esp32/capstone/src/main.cpp

```diff
@@ -0,0 +1,283 @@
+#include <Arduino.h>
+#include <FastLED.h>
+#include <FreeRTOS.h>
+#include "soc/rtc_wdt.h"
+
+#define DEMO 1 /* Demo displays a preset image */
+#define DEBUG
+
+#ifdef DEBUG
+#define DEBUG_PRINT(x) Serial.println(x)
+#else
+#define DEBUG_PRINT(x)
+#endif
+
+const int NUM_ANGULAR_SEGMENTS = 120; // 3 degree segments
+const int NUM_LEDS = 140; // 144-4
+const int NUM_LEDS_ARM = 35; // 140 leds total/4 arms
+
+const int LED_DATA_PIN = 15; //gpio #15
+const int HES_SIG_PIN = A2; // gpi #34
+
+const int MOTOR_RPM_THRESHOLD = 100; // images won't display while below this threshold
+
+// define a few colors for testing
+#define BLUE    CRGB::Blue
+#define RED     CRGB::Red
+#define GREEN   CRGB::Green
+
+typedef struct ImageFrameData
+{
+    // array of RGB values for LEDs
+    CRGB ledValues[NUM_LEDS_ARM][NUM_ANGULAR_SEGMENTS];
+} ImageFrameData;
+
+ImageFrameData *demoImage;
+
+float g_currentRpm;
+volatile boolean g_newRotation;
+
+// used for calculating RPM
+volatile unsigned int g_revolutions;
+unsigned long g_prevTimeMs;
+float g_revsPerMs;
+float g_timePer90degMs; // time for arms to rotate 90deg
+float g_timePerSegment_us; // time for arms to rotate through 1 angular segment
+
+hw_timer_t *quarter_circle_timer = NULL;
+
+// arm-quarter mappings (should shift every 90deg)
+volatile int g_arm_quarter_map[4];
+
+portMUX_TYPE timerMux = portMUX_INITIALIZER_UNLOCKED;
+portMUX_TYPE hesMux = portMUX_INITIALIZER_UNLOCKED;
+
+// led values array for FastLED
+CRGB g_leds[NUM_LEDS];
+
+/* FreeRTOS things */
+TaskHandle_t ImageDisplayTask;
+TaskHandle_t CommunicationTask;
+QueueHandle_t ImageFrameQueue;
+
+
+/* Hall effect sensor interrupt service routine */
+void IRAM_ATTR HallEffectSensor_ISR()
+{
+    portENTER_CRITICAL_ISR(&hesMux);
+    g_newRotation = true;
+    g_revolutions++;
+    portEXIT_CRITICAL_ISR(&hesMux);
+}
+
+/* ISR for HW timer */
+void IRAM_ATTR Timer90Deg_ISR()
+{
+    portENTER_CRITICAL_ISR(&timerMux);
+    for (int i = 0; i < 4; i++)
+    {
+        if (g_arm_quarter_map[i] == 3)
+        {            
+            g_arm_quarter_map[i] =0;
+        }
+        else
+        {
+            g_arm_quarter_map[i]++;
+        } 
+    }
+    portEXIT_CRITICAL_ISR(&timerMux);
+}
+
+
+void CalculateRPM()
+{
+    // do new calculation every 3 revolutions - decrease for improved accuracy
+    if (g_revolutions >= 3)
+    {
+        if (g_prevTimeMs != 0) // dont divide by 0
+            g_revsPerMs = g_revolutions/((float)(millis() - g_prevTimeMs));
+        g_prevTimeMs = millis();
+        g_revolutions = 0;
```

Comment:
I'm not sure how expensive a call to `millis()` is but we could cache the result of a single `millis()` call into a local variable here to avoid calling it multiple times (this might also help give us slightly more accurate RPM calculations) i.e.
```C
unsigned long currentTimeMillis = millis();

if (g_prevTimeMs != 0) // dont divide by 0
            g_revsPerMs = g_revolutions/((float)(currentTimeMillis - g_prevTimeMs));
g_prevTimeMs = currentTimeMillis;
```